Numerical constant errors and improvements

bbbradsmith opened the pull request:
This is a few improvements to do with numerical constants.

* Added an error for numerical constants that overflow the 32-bit internal limitation. See: #2026
* Added optional warning ```const-overflow``` for a numerical constant that overflows its storage, off by default. (I believe this is a default warning for clang, but "pedantic" for GCC.)
* Fix incomplete/broken parsing of floating point constants, also allows unary +/- for negative floating point.
* Allow floating point constants to be implicitly converted to integers, with a warning if precision is lost.
* Better error message when unsupported floating point arithmetic is attempted.
* Documenation and simple test of floating point constants.

The floating point stuff was addressing #2021 which revealed that at some point partial support for floating point constants was attempted. However, it was both incomplete and broken, and instead of producing any error it just silently turned them all into 0 valued integers. I've tried to complete the implementation of their parsing, and allow simple conversion to integers. The warning should avoid any surprise about the precision loss, and the lack of expression arithmetic support now has a more intuitive error. I also added a small test of the feature and added a note about it to the documentation.

bbbradsmith commented:
As a side note, if someone wanted to take floating point constant support a bit further, and allow arithmetic in their expressions, promoting int to float, etc. [datatype.c:ArithmeticConvert](https://github.com/cc65/cc65/blob/8f2927cf07d92966c9f61b3e0d65e2509b382d98/src/cc65/datatype.c#L660) appears to be the entry point for this.

mrdudz commented:
Would you please split this PR and put the floating point stuff into a seperate one? One of my side quests is actually fixing the float support, and that would perhaps make it a bit easier for me (see https://github.com/mrdudz/cc65/tree/fptest )

bbbradsmith commented:
As requested I have split and rebased this into a pair of pull requests: #2084 and #2085 

The individual commits were each a separate functional chunk, so even though both of these commits have multiple changes, looking at individual commits should show each isolated component change, if you need to investigate or be selective.

mrdudz commented:
thanks, closing this one then
